drivers/netdev_common/netdev_common.c

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright (C) 2020 Benjamin Valentin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_netdev_ieee802154
+ * @{
+ *
+ * @file
+ * @author  Benjamin Valentin <benjamin.valentin@ml-pa-com>
+ */
+
+#include "net/netdev.h"
+
+void netdev_common_init(struct netdev *dev)
+{
+#if IS_ACTIVE(GNRC_NETIF_SINGLE)
+    (void) dev;
+#else
+    static netdev_id_t id;
+    dev->id = id++;
```

Comment:
IMHO this should not be dependent on initialization order. An ID that is the same across all compile configurations would be the best IMHO. One idea that comes to mind would be a pair of `uint8_t` the first one representing the device model, the second e.g. representative of the configuration. Both could be some kind of hashes/sum (for the device model e.g. of the name, for the parameters of the param struct).

Reply:
I've updated (rewritten) the PR to match this.

Now before I update every `setup()` function for every driver, I provided an example for `at86rf2xx`.